archive/untar_stream.ts

```diff
@@ -0,0 +1,235 @@
+ *   These archives typically have a single '.tar' extension.  This
+ * implementation follows the [FreeBSD 15.0](https://man.freebsd.org/cgi/man.cgi?query=tar&sektion=5&apropos=0&manpath=FreeBSD+15.0-CURRENT) spec.
+ *
+ * ### Supported File Formats
+ * Only the ustar file format is supported.  This is the most common format.
+ *   Additionally the numeric extension for file size.
+ *
+ * ### Usage
+ * When expanding the archive, as demonstrated in the example, one must decide
+ * to either consume the Readable Stream, if present, or cancel it. The next
+ * entry won't be resolved until the previous ReadableStream is either consumed
+ * or cancelled.
+ *
+ * ### Understanding Compressed
+ * A tar archive may be compressed, often identified by an additional file
+ * extension, such as '.tar.gz' for gzip. This TransformStream does not support
+ * decompression which must be done before expanding the archive.
+ *
+ * @example
+ * ```ts
+ * import { UnTarStream } from "@std/archive/untar-stream";
+ *
+ * let fileWriter: WritableStreamDefaultWriter | undefined;
+ * for await (
+ *   const entry of (await Deno.open('./out.tar.gz'))
+ *     .readable
+ *     .pipeThrough(new DecompressionStream('gzip'))
+ *     .pipeThrough(new UnTarStream())
+ * ) {
+ *   if (entry.type === "header") {
+ *     fileWriter?.close();
+ *     fileWriter = (await Deno.create(entry.pathname)).writable.getWriter();
+ *   } else await fileWriter!.write(entry.data);
+ * }
+ * ```
+ */
+export class UnTarStream
+  implements TransformStream<Uint8Array, TarStreamChunk> {
+  #readable: ReadableStream<TarStreamChunk>;
+  #writable: WritableStream<Uint8Array>;
+  #gen: AsyncGenerator<Uint8Array>;
+  constructor() {
+    const { readable, writable } = new TransformStream<
+      Uint8Array,
+      Uint8Array
+    >();
+    this.#readable = ReadableStream.from(this.#untar());
+    this.#writable = writable;
+
+    this.#gen = async function* () {
+      const buffer: Uint8Array[] = [];
+      for await (
+        const chunk of readable.pipeThrough(new FixedChunkStream(512))
+      ) {
+        if (chunk.length !== 512) {
+          throw new Error("Tarball has an unexpected number of bytes");
+        }
+
+        buffer.push(chunk);
+        if (buffer.length > 2) yield buffer.shift()!;
+      }
+      if (buffer.length < 2) {
+        throw new Error("Tarball was too small to be valid");
+      }
+      if (!buffer.every((value) => value.every((x) => x === 0))) {
+        throw new Error("Tarball has invalid ending");
+      }
+    }();
+  }
+
+  async *#untar(): AsyncGenerator<TarStreamChunk> {
+    const decoder = new TextDecoder();
+    while (true) {
+      const { done, value } = await this.#gen.next();
+      if (done) break;
+
+      // Validate Checksum
+      const checksum = value.slice();
+      checksum.set(new Uint8Array(8).fill(32), 148);
+      if (
+        checksum.reduce((x, y) => x + y) !==
+          parseInt(decoder.decode(value.slice(148, 156 - 2)), 8)
```

Comment:
To avoid copying all 512 bytes:

```suggestion
      const checksum = parseInt(decoder.decode(value.subarray(148, 156)), 8);
      value.fill(32, 148, 156);
      if (
        value.reduce((x, y) => x + y) !== checksum
```